File: src/main/java/com/task/twinotask/service/ClientService.java
package com.task.twinotask.service;

import com.task.twinotask.entity.Client;
import com.task.twinotask.entity.ProfileVisibility;
import com.task.twinotask.entity.Role;
import com.task.twinotask.exceptions.UserAlreadyExistException;
import com.task.twinotask.repository.ClientRepository;
import com.task.twinotask.web.dto.ClientRegistrationDto;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.User;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

@Service
public class ClientService implements UserDetailsService {

	private ClientRepository clientRepository;

	private BCryptPasswordEncoder passwordEncoder;

	@Autowired
	public void setClientRepository(ClientRepository repository) {
		clientRepository = repository;
	}

	@Autowired
	public void setPasswordEncoder(BCryptPasswordEncoder passwordEncoder) {
		this.passwordEncoder = passwordEncoder;
	}

	@SuppressWarnings("UnusedReturnValue")
	public Client registerClient(final ClientRegistrationDto registration) throws UserAlreadyExistException {
		if (userExists(registration.getEmail())) {
			throw new UserAlreadyExistException("There is an account with that email address: " + registration.getEmail());
		}

		Client client = new Client(
				Objects.requireNonNull(registration.getFirstName()),
				Objects.requireNonNull(registration.getLastName()),
				Objects.requireNonNull(registration.getEmail()),
				passwordEncoder.encode(Objects.requireNonNull(registration.getPassword())),
				registration.getPhoneNumber(),
				registration.getBirthDate(),
				registration.getSalary(),
				registration.getLiabilities(),
				Collections.singletonList(new Role("ROLE_USER", -1)),
				ProfileVisibility.REGISTERED,
				false,
				0L
		);

		return clientRepository.save(client);
	}

	@Override
	public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
		Client client = clientRepository.findByEmail(email);
		if (client == null) {
			throw new UsernameNotFoundException("Invalid username.");
		}
		return new User(client.getEmail(),
				client.getPassword(),
				mapRolesToAuthorities(Objects.requireNonNull(client.getRoles())));
	}

	private Collection<? extends GrantedAuthority> mapRolesToAuthorities(Collection<Role> roles) {
		return roles.stream()
				.map(role -> new SimpleGrantedAuthority(role.getName()))
				.collect(Collectors.toList());
	}

	public List<Client> findAll() {
		return clientRepository.findAll();
	}

	public Client findById(final Long id) {
		return clientRepository.findOne(id);
	}

	public Client findByEmail(final String email) {
		return clientRepository.findByEmail(email);
	}

	@SuppressWarnings("WeakerAccess")
	public boolean userExists(final String email) {
		return clientRepository.findByEmail(email) != null;
	}

	@SuppressWarnings("unused")
	public void delete(final Client user) {
		clientRepository.delete(user);
	}

}

File: src/main/java/com/task/twinotask/service/ClientService.kt
package com.task.twinotask.service

import com.task.twinotask.entity.Client
import com.task.twinotask.entity.Role
import com.task.twinotask.exceptions.UserAlreadyExistException
import com.task.twinotask.repository.ClientRepository
import com.task.twinotask.web.dto.ClientRegistrationDto
import org.springframework.context.annotation.Lazy
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.User
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.stereotype.Service

@Service
class ClientService(
	private val clientRepository: ClientRepository,
	@Lazy private var passwordEncoder: BCryptPasswordEncoder
) : UserDetailsService {

	fun registerClient(registration: ClientRegistrationDto): Client {
		if (userExists(registration.email)) {
			throw UserAlreadyExistException(
				"There is an account with that email address: " + registration.email
			)
		}

		val client = Client(
			registration.firstName,
			registration.lastName,
			registration.email,
			passwordEncoder.encode(registration.password),
			registration.phoneNumber,
			registration.birthDate,
			registration.salary,
			registration.liabilities,
			listOf(Role("ROLE_USER"))
		)

		return clientRepository.save(client)
	}

	override fun loadUserByUsername(email: String): UserDetails {
		val (_, _, userEmail, password, _, _, _, _, roles) = clientRepository.findByEmail(email)
				?: throw UsernameNotFoundException("Invalid username.")
		return User(
			userEmail,
			password,
			mapRolesToAuthorities(roles!!)
		)
	}

	private fun mapRolesToAuthorities(roles: Collection<Role>): Collection<GrantedAuthority> =
		roles.map { (name) -> SimpleGrantedAuthority(name) }.toList()

	fun findAll(): MutableList<Client> = clientRepository.findAll()

	fun findById(id: Long?): Client? = clientRepository.findOne(id)

	fun findByEmail(email: String): Client? = clientRepository.findByEmail(email)

	fun userExists(email: String) = clientRepository.findByEmail(email) != null

	@Suppress("unused")
	fun delete(user: Client) {
		clientRepository.delete(user)
	}

}
